File: ios/Runner/Engine/CameraFilterEnginePlugin.swift
import Flutter
import UIKit

public class CameraFilterEnginePlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private weak var registrar: FlutterPluginRegistrar?
    private var engine: FilterEngine?
    private var progressSink: FlutterEventSink?
    private var currentVideoCancel: MediaProcessor.CancelFlag?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "camera_filter_engine",
            binaryMessenger: registrar.messenger()
        )
        let instance = CameraFilterEnginePlugin()
        instance.registrar = registrar
        registrar.addMethodCallDelegate(instance, channel: channel)

        let events = FlutterEventChannel(
            name: "camera_filter_engine/progress",
            binaryMessenger: registrar.messenger()
        )
        events.setStreamHandler(instance)
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        progressSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        progressSink = nil
        return nil
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        switch call.method {
        case "initialize":
            initialize(args: args, result: result)
        case "dispose":
            engine?.shutdown()
            engine = nil
            result(nil)
        case "setFilter":
            let id = args["id"] as? String ?? "none"
            let params = (args["params"] as? [String: NSNumber])?
                .mapValues { $0.floatValue }
            engine?.setFilter(id: id, params: params)
            result(nil)
        case "setParam":
            if let k = args["key"] as? String,
               let v = (args["value"] as? NSNumber)?.floatValue {
                engine?.setParam(key: k, value: v)
            }
            result(nil)
        case "setLut":
            engine?.setLut(path: args["path"] as? String)
            result(nil)
        case "switchCamera":
            let front = engine?.switchCamera() ?? false
            result(front ? "front" : "back")
        case "setZoom":
            let level = (args["level"] as? NSNumber)?.floatValue ?? 0
            engine?.setZoom(level: level)
            result(nil)
        case "takePicture":
            if let path = args["path"] as? String {
                engine?.takePicture(path: path) { saved in result(saved) }
            } else {
                result(FlutterError(code: "ARG", message: "path required", details: nil))
            }
        case "startRecording":
            if let path = args["path"] as? String {
                engine?.startRecording(path: path)
            }
            result(nil)
        case "stopRecording":
            engine?.stopRecording { path in result(path) }
        case "saveToGallery":
            let path = args["path"] as? String ?? ""
            let isVideo = (args["isVideo"] as? Bool) ?? false
            GallerySaver.save(path: path, isVideo: isVideo) { ok in
                DispatchQueue.main.async { result(ok) }
            }
        case "processImage":
            processImage(args: args, result: result)
        case "processVideo":
            processVideo(args: args, result: result)
        case "previewFilter":
            previewFilter(args: args, result: result)
        case "cropImage":
            cropImage(args: args, result: result)
        case "cropVideo":
            cropVideo(args: args, result: result)
        case "composeVideo":
            composeVideo(args: args, result: result)
        case "trimVideo":
            trimVideo(args: args, result: result)
        case "cancelProcessing":
            currentVideoCancel?.cancelled = true
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func initialize(args: [String: Any], result: @escaping FlutterResult) {
        guard let registrar = registrar else {
            result(FlutterError(code: "STATE", message: "registrar missing", details: nil))
            return
        }
        let width  = (args["width"]  as? Int) ?? 1280
        let height = (args["height"] as? Int) ?? 720
        let lens   = (args["lens"]   as? String) ?? "back"
        do {
            let e = try FilterEngine(
                textureRegistry: registrar.textures(),
                width: width,
                height: height,
                startFront: lens == "front"
            )
            try e.start()
            engine = e
            result([
                "textureId": e.textureId,
                "width": width,
                "height": height,
            ])
        } catch {
            result(FlutterError(code: "INIT", message: error.localizedDescription, details: nil))
        }
    }

    private func processImage(args: [String: Any], result: @escaping FlutterResult) {
        guard let input = args["inputPath"] as? String,
              let output = args["outputPath"] as? String,
              let filterId = args["filterId"] as? String else {
            result(FlutterError(code: "ARG", message: "missing args", details: nil))
            return
        }
        let params = (args["params"] as? [String: NSNumber])?.mapValues { $0.floatValue }
        let lut = args["lutPath"] as? String
        let filterId2 = args["filterId2"] as? String
        let params2 = (args["params2"] as? [String: NSNumber])?.mapValues { $0.floatValue }
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let path = try MediaProcessor.processImage(
                    inputPath: input,
                    outputPath: output,
                    filterId: filterId,
                    params: params,
                    lutPath: lut,
                    filterId2: filterId2,
                    params2: params2
                )
                DispatchQueue.main.async { result(path) }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "PROC", message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    private func previewFilter(args: [String: Any], result: @escaping FlutterResult) {
        guard let input = args["inputPath"] as? String,
              let output = args["outputPath"] as? String,
              let filterId = args["filterId"] as? String else {
            result(FlutterError(code: "ARG", message: "missing args", details: nil))
            return
        }
        let params = (args["params"] as? [String: NSNumber])?.mapValues { $0.floatValue }
        let lut = args["lutPath"] as? String
        let isVideo = (args["isVideo"] as? Bool) ?? false
        let atSeconds = (args["atSeconds"] as? NSNumber)?.doubleValue ?? 1.0
        let filterId2 = args["filterId2"] as? String
        let params2 = (args["params2"] as? [String: NSNumber])?.mapValues { $0.floatValue }
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let path = try MediaProcessor.previewFilter(
                    inputPath: input, outputPath: output,
                    filterId: filterId, params: params, lutPath: lut,
                    isVideo: isVideo, atSeconds: atSeconds,
                    filterId2: filterId2, params2: params2
                )
                DispatchQueue.main.async { result(path) }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "PROC", message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    private func cropImage(args: [String: Any], result: @escaping FlutterResult) {
        guard let input = args["inputPath"] as? String,
              let output = args["outputPath"] as? String else {
            result(FlutterError(code: "ARG", message: "missing args", details: nil))
            return
        }
        let l = (args["left"] as? NSNumber)?.doubleValue ?? 0.0
        let t = (args["top"] as? NSNumber)?.doubleValue ?? 0.0
        let r = (args["right"] as? NSNumber)?.doubleValue ?? 1.0
        let b = (args["bottom"] as? NSNumber)?.doubleValue ?? 1.0
        let flipH = (args["flipH"] as? Bool) ?? false
        let flipV = (args["flipV"] as? Bool) ?? false
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let path = try MediaProcessor.cropImage(
                    inputPath: input, outputPath: output,
                    left: l, top: t, right: r, bottom: b,
                    flipH: flipH, flipV: flipV
                )
                DispatchQueue.main.async { result(path) }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "CROP", message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    private func composeVideo(args: [String: Any], result: @escaping FlutterResult) {
        guard let input = args["inputPath"] as? String,
              let output = args["outputPath"] as? String,
              let overlay = args["overlayPngPath"] as? String else {
            result(FlutterError(code: "ARG", message: "missing args", details: nil))
            return
        }
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let path = try MediaProcessor.composeVideo(
                    inputPath: input, outputPath: output,
                    overlayPngPath: overlay
                )
                DispatchQueue.main.async { result(path) }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "COMPOSE", message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    private func cropVideo(args: [String: Any], result: @escaping FlutterResult) {
        guard let input = args["inputPath"] as? String,
              let output = args["outputPath"] as? String else {
            result(FlutterError(code: "ARG", message: "missing args", details: nil))
            return
        }
        let l = (args["left"] as? NSNumber)?.doubleValue ?? 0.0
        let t = (args["top"] as? NSNumber)?.doubleValue ?? 0.0
        let r = (args["right"] as? NSNumber)?.doubleValue ?? 1.0
        let b = (args["bottom"] as? NSNumber)?.doubleValue ?? 1.0
        let flipH = (args["flipH"] as? Bool) ?? false
        let flipV = (args["flipV"] as? Bool) ?? false
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let path = try MediaProcessor.cropVideo(
                    inputPath: input, outputPath: output,
                    left: l, top: t, right: r, bottom: b,
                    flipH: flipH, flipV: flipV
                )
                DispatchQueue.main.async { result(path) }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "CROP", message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    private func trimVideo(args: [String: Any], result: @escaping FlutterResult) {
        guard let input = args["inputPath"] as? String,
              let output = args["outputPath"] as? String else {
            result(FlutterError(code: "ARG", message: "missing args", details: nil))
            return
        }
        let startMs = (args["startMs"] as? NSNumber)?.int64Value ?? 0
        let endMs = (args["endMs"] as? NSNumber)?.int64Value ?? Int64.max
        DispatchQueue.global(qos: .userInitiated).async {
            do {
                let path = try MediaProcessor.trimVideo(
                    inputPath: input, outputPath: output,
                    startMs: startMs, endMs: endMs
                )
                DispatchQueue.main.async { result(path) }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "TRIM", message: error.localizedDescription, details: nil))
                }
            }
        }
    }

    private func processVideo(args: [String: Any], result: @escaping FlutterResult) {
        guard let input = args["inputPath"] as? String,
              let output = args["outputPath"] as? String,
              let filterId = args["filterId"] as? String else {
            result(FlutterError(code: "ARG", message: "missing args", details: nil))
            return
        }
        let params = (args["params"] as? [String: NSNumber])?.mapValues { $0.floatValue }
        let lut = args["lutPath"] as? String
        let filterId2 = args["filterId2"] as? String
        let params2 = (args["params2"] as? [String: NSNumber])?.mapValues { $0.floatValue }

        // Tell any in-flight processVideo to stop, then own the new flag.
        currentVideoCancel?.cancelled = true
        let cancel = MediaProcessor.CancelFlag()
        currentVideoCancel = cancel

        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            defer {
                if self?.currentVideoCancel === cancel { self?.currentVideoCancel = nil }
            }
            do {
                let path = try MediaProcessor.processVideo(
                    inputPath: input,
                    outputPath: output,
                    filterId: filterId,
                    params: params,
                    lutPath: lut,
                    progress: { p in
                        self?.progressSink?(p)
                    },
                    cancel: cancel,
                    filterId2: filterId2,
                    params2: params2
                )
                DispatchQueue.main.async { result(path) }
            } catch is MediaProcessor.CancelledError {
                // Cancelled — return nil so Dart can ignore quietly.
                DispatchQueue.main.async { result(nil) }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "PROC", message: error.localizedDescription, details: nil))
                }
            }
        }
    }
}

File: android/app/src/main/kotlin/com/example/camera_filter_engine/engine/CameraFilterEnginePlugin.kt
package com.example.camera_filter_engine.engine

import android.app.Activity
import android.content.Context
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.view.TextureRegistry
import android.os.Handler
import android.os.Looper
import java.util.concurrent.Executors

class CameraFilterEnginePlugin(private val initialActivity: Activity? = null)
    : FlutterPlugin, ActivityAware, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

    private lateinit var channel: MethodChannel
    private lateinit var progressChannel: EventChannel
    private lateinit var context: Context
    private lateinit var textures: TextureRegistry
    private var activity: Activity? = initialActivity
    private var lifecycle: Lifecycle? = null

    private var engine: FilterEngine? = null
    private var progressSink: EventChannel.EventSink? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    private val ioExecutor = Executors.newSingleThreadExecutor()
    // Flag for the currently running processVideo job; flipping it to true
    // tells the GL/MediaCodec loop to bail out early.
    private var currentVideoCancel: java.util.concurrent.atomic.AtomicBoolean? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        textures = binding.textureRegistry
        channel = MethodChannel(binding.binaryMessenger, "camera_filter_engine")
        channel.setMethodCallHandler(this)
        progressChannel = EventChannel(binding.binaryMessenger, "camera_filter_engine/progress")
        progressChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        progressChannel.setStreamHandler(null)
        engine?.release()
        engine = null
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        progressSink = events
    }

    override fun onCancel(arguments: Any?) {
        progressSink = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        lifecycle = (binding.activity as? LifecycleOwner)?.lifecycle
    }
    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) =
        onAttachedToActivity(binding)
    override fun onDetachedFromActivityForConfigChanges() { activity = null }
    override fun onDetachedFromActivity() { activity = null }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        try {
            when (call.method) {
                "initialize" -> initialize(call, result)
                "dispose" -> { engine?.release(); engine = null; result.success(null) }
                "setFilter" -> {
                    val id = call.argument<String>("id") ?: "none"
                    @Suppress("UNCHECKED_CAST")
                    val params = call.argument<Map<String, Any?>>("params")
                            ?.mapValues { (it.value as Number).toFloat() }
                    engine?.setFilter(id, params)
                    result.success(null)
                }
                "setParam" -> {
                    val k = call.argument<String>("key")!!
                    val v = (call.argument<Number>("value") ?: 0.0).toFloat()
                    engine?.setParam(k, v)
                    result.success(null)
                }
                "setLut" -> {
                    val path = call.argument<String>("path")
                    engine?.setLut(path)
                    result.success(null)
                }
                "switchCamera" -> {
                    val front = engine?.switchCamera() ?: false
                    result.success(if (front) "front" else "back")
                }
                "setZoom" -> {
                    val level = (call.argument<Number>("level") ?: 0.0).toFloat()
                    engine?.setZoom(level)
                    result.success(null)
                }
                "takePicture" -> {
                    val path = call.argument<String>("path")!!
                    engine?.takePicture(path) { p -> result.success(p) }
                        ?: result.success(null)
                }
                "startRecording" -> {
                    val path = call.argument<String>("path")!!
                    engine?.startRecording(path)
                    result.success(null)
                }
                "stopRecording" -> {
                    val path = engine?.stopRecording()
                    result.success(path)
                }
                "saveToGallery" -> {
                    val path = call.argument<String>("path")!!
                    val isVideo = call.argument<Boolean>("isVideo") ?: false
                    val ok = GallerySaver.save(context, path, isVideo)
                    result.success(ok)
                }
                "processImage" -> handleProcessImage(call, result)
                "processVideo" -> handleProcessVideo(call, result)
                "previewFilter" -> handlePreviewFilter(call, result)
                "cropImage" -> handleCropImage(call, result)
                "cropVideo" -> handleCropVideo(call, result)
                "composeVideo" -> handleComposeVideo(call, result)
                "trimVideo" -> handleTrimVideo(call, result)
                "cancelProcessing" -> {
                    currentVideoCancel?.set(true)
                    result.success(null)
                }
                else -> result.notImplemented()
            }
        } catch (e: Exception) {
            result.error("CFE_ERROR", e.message, e.stackTraceToString())
        }
    }

    private fun initialize(call: MethodCall, result: MethodChannel.Result) {
        val act = activity ?: throw IllegalStateException("Activity not attached")
        val width = call.argument<Int>("width") ?: 1280
        val height = call.argument<Int>("height") ?: 720
        val lensStr = call.argument<String>("lens") ?: "back"

        val entry = textures.createSurfaceTexture()
        val e = FilterEngine(
            context = context,
            activity = act,
            lifecycle = lifecycle,
            textureEntry = entry,
            targetWidth = width,
            targetHeight = height,
            startFront = lensStr == "front",
        )
        e.start()
        engine = e

        result.success(mapOf(
            "textureId" to entry.id(),
            "width" to width,
            "height" to height,
        ))
    }

    private fun handleProcessImage(call: MethodCall, result: MethodChannel.Result) {
        val input = call.argument<String>("inputPath")!!
        val output = call.argument<String>("outputPath")!!
        val filterId = call.argument<String>("filterId") ?: "none"
        val params = floatParams(call, "params")
        val lutPath = call.argument<String>("lutPath")
        val filterId2 = call.argument<String>("filterId2")
        val params2 = floatParams(call, "params2")
        ioExecutor.execute {
            try {
                val saved = MediaProcessor.processImage(context, input, output, filterId, params,
                    lutPath, filterId2, params2)
                mainHandler.post { result.success(saved) }
            } catch (t: Throwable) {
                mainHandler.post { result.error("PROC", t.message, t.stackTraceToString()) }
            }
        }
    }

    private fun handleCropImage(call: MethodCall, result: MethodChannel.Result) {
        val input = call.argument<String>("inputPath")!!
        val output = call.argument<String>("outputPath")!!
        val left = (call.argument<Number>("left") ?: 0.0).toDouble()
        val top = (call.argument<Number>("top") ?: 0.0).toDouble()
        val right = (call.argument<Number>("right") ?: 1.0).toDouble()
        val bottom = (call.argument<Number>("bottom") ?: 1.0).toDouble()
        val flipH = call.argument<Boolean>("flipH") ?: false
        val flipV = call.argument<Boolean>("flipV") ?: false
        ioExecutor.execute {
            try {
                val saved = MediaProcessor.cropImage(context, input, output,
                    left, top, right, bottom, flipH, flipV)
                mainHandler.post { result.success(saved) }
            } catch (t: Throwable) {
                mainHandler.post { result.error("CROP", t.message, t.stackTraceToString()) }
            }
        }
    }

    private fun handleCropVideo(call: MethodCall, result: MethodChannel.Result) {
        val input = call.argument<String>("inputPath")!!
        val output = call.argument<String>("outputPath")!!
        val left = (call.argument<Number>("left") ?: 0.0).toDouble()
        val top = (call.argument<Number>("top") ?: 0.0).toDouble()
        val right = (call.argument<Number>("right") ?: 1.0).toDouble()
        val bottom = (call.argument<Number>("bottom") ?: 1.0).toDouble()
        val flipH = call.argument<Boolean>("flipH") ?: false
        val flipV = call.argument<Boolean>("flipV") ?: false
        ioExecutor.execute {
            try {
                val saved = MediaProcessor.cropVideo(context, input, output,
                    left, top, right, bottom, flipH, flipV)
                mainHandler.post { result.success(saved) }
            } catch (t: Throwable) {
                mainHandler.post { result.error("CROP", t.message, t.stackTraceToString()) }
            }
        }
    }

    private fun handleComposeVideo(call: MethodCall, result: MethodChannel.Result) {
        val input = call.argument<String>("inputPath")!!
        val output = call.argument<String>("outputPath")!!
        val overlay = call.argument<String>("overlayPngPath")!!
        ioExecutor.execute {
            try {
                val saved = MediaProcessor.composeVideo(context, input, output, overlay)
                mainHandler.post { result.success(saved) }
            } catch (t: Throwable) {
                mainHandler.post { result.error("COMPOSE", t.message, t.stackTraceToString()) }
            }
        }
    }

    private fun handleTrimVideo(call: MethodCall, result: MethodChannel.Result) {
        val input = call.argument<String>("inputPath")!!
        val output = call.argument<String>("outputPath")!!
        val startMs = (call.argument<Number>("startMs") ?: 0).toLong()
        val endMs = (call.argument<Number>("endMs") ?: Long.MAX_VALUE).toLong()
        ioExecutor.execute {
            try {
                val saved = MediaProcessor.trimVideo(input, output, startMs, endMs)
                mainHandler.post { result.success(saved) }
            } catch (t: Throwable) {
                mainHandler.post { result.error("TRIM", t.message, t.stackTraceToString()) }
            }
        }
    }

    private fun floatParams(call: MethodCall, key: String): Map<String, Float>? {
        @Suppress("UNCHECKED_CAST")
        return call.argument<Map<String, Any?>>(key)
            ?.mapValues { (it.value as Number).toFloat() }
    }

    private fun handlePreviewFilter(call: MethodCall, result: MethodChannel.Result) {
        val input = call.argument<String>("inputPath")!!
        val output = call.argument<String>("outputPath")!!
        val filterId = call.argument<String>("filterId") ?: "none"
        val isVideo = call.argument<Boolean>("isVideo") ?: false
        val atSeconds = (call.argument<Number>("atSeconds") ?: 1.0).toDouble()
        val params = floatParams(call, "params")
        val lutPath = call.argument<String>("lutPath")
        val filterId2 = call.argument<String>("filterId2")
        val params2 = floatParams(call, "params2")
        ioExecutor.execute {
            try {
                val saved = MediaProcessor.previewFilter(
                    context, input, output, filterId, params, lutPath,
                    isVideo, atSeconds, filterId2, params2,
                )
                mainHandler.post { result.success(saved) }
            } catch (t: Throwable) {
                mainHandler.post { result.error("PROC", t.message, t.stackTraceToString()) }
            }
        }
    }

    private fun handleProcessVideo(call: MethodCall, result: MethodChannel.Result) {
        val input = call.argument<String>("inputPath")!!
        val output = call.argument<String>("outputPath")!!
        val filterId = call.argument<String>("filterId") ?: "none"
        val params = floatParams(call, "params")
        val lutPath = call.argument<String>("lutPath")
        val filterId2 = call.argument<String>("filterId2")
        val params2 = floatParams(call, "params2")

        // Tell any in-flight processVideo job to stop, then own the new flag.
        currentVideoCancel?.set(true)
        val cancel = java.util.concurrent.atomic.AtomicBoolean(false)
        currentVideoCancel = cancel

        ioExecutor.execute {
            try {
                val saved = MediaProcessor.processVideo(
                    context, input, output, filterId, params, lutPath,
                    progress = { p ->
                        val pd = p.toDouble()
                        mainHandler.post { progressSink?.success(pd) }
                    },
                    cancel = cancel,
                    filterId2 = filterId2,
                    params2 = params2,
                )
                mainHandler.post { result.success(saved) }
            } catch (_: MediaProcessor.CancelledException) {
                // Cancelled — return null path so Dart can ignore quietly.
                mainHandler.post { result.success(null) }
            } catch (t: Throwable) {
                mainHandler.post { result.error("PROC", t.message, t.stackTraceToString()) }
            } finally {
                // Drop the flag if it's still the one we owned; new callers
                // will have replaced it already if they raced in.
                if (currentVideoCancel === cancel) currentVideoCancel = null
            }
        }
    }
}
